hwtracer/src/decode/ykpt/mod.rs

```diff
@@ -97,126 +280,355 @@ impl<'t> YkPTBlockIterator<'t> {
+                        reposition = true;
+                    }
+                }
+                iced_x86::FlowControl::UnconditionalBranch => {
+                    dis.set_ip(self.branch_target_vaddr(&inst));
+                    reposition = true;
                 }
-            } else {
-                // In the absence of blockmap info, we cannot know the successor block.
-                self.cur_loc = ObjLoc::OtherObjOrUnknown;
-                return Ok(Some(Block::unknown()));
+                iced_x86::FlowControl::Call => {
+                    if inst.code() == iced_x86::Code::Syscall {
+                        // Do nothing. We assume the kernel isn't traced.
+                    } else {
+                        let target_vaddr = self.branch_target_vaddr(&inst);
+                        if target_vaddr == inst.next_ip() {
+                            // FIXME: Intel PT treats this case specially.
+                            todo!();
+                        }
+                        // We don't expect to see any 16-bit mode far calls in modernity.
+                        debug_assert!(!inst.is_call_far());
+                        self.comprets
+                            .push(CompRetAddr::VAddr(usize::try_from(inst.next_ip()).unwrap()));
+                        dis.set_ip(target_vaddr);
+                        reposition = true;
+                    }
+                }
+                _ => todo!(),
             }
         }
+    }
 
-        // If we get here then we can't statically figure out where to go next. A packet is going
-        // to have to tell us what to do.
-        self.seek_tip()
+    /// Skip over "foreign code" for which we have no blockmap info for.
+    fn skip_foreign(&mut self, start_vaddr: Option<usize>) -> Result<Block, HWTracerError> {
+        let start_vaddr = match start_vaddr {
+            Some(v) => v,
+            None => {
+                // We don't statically know where to start, so we rely on a TIP update to tell us.
+                self.seek_tip()?;
+                match self.cur_loc {
+                    ObjLoc::OtherObjOrUnknown(Some(vaddr)) => vaddr,
+                    ObjLoc::OtherObjOrUnknown(None) => {
+                        // The above `seek_tip()` ensures this can't happen!
+                        unreachable!();
+                    }
+                    ObjLoc::MainObj(off) => self.off_to_vaddr(&PHDR_MAIN_OBJ, off)?,
+                }
+            }
+        };
+        self.disassemble(start_vaddr)
     }
 
-    /// Keep decoding packets until we have some TNT decisions buffered.
-    fn seek_tnt(&mut self) -> Result<bool, HWTracerError> {
-        while self.tnts.is_empty() {
+    /// Keep decoding packets until we encounter a TNT packet.
+    fn seek_tnt(&mut self) -> Result<(), HWTracerError> {
+        loop {
             let pkt = self.packet()?; // Potentially populates `self.tnts`.
```

Comment:
Does this return `Err` if empty? If not, could we get an infinite loop?

Reply:
If the trace is exhausted it returns the new `HWTracer::NoMorePackets` error.

Reply:
Got it!